### PR TITLE
WIP: enforcing certificate validation for Envoy

### PR DESCRIPTION
**Pull request by alyssawilk**
I'm about halfway through the many annoying text fix-ups when it occurred to me I want to get an API review before I do the rest of the fixes.

specifically given there's TrustChainVerification ACCEPT_UNTRUSTED I wonder if we should use that for enforcement rather than adding a new boolean.  And if we want a boolean if we should name it skip_verification_for_test or if the current name is good or too snarky.

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
part of https://github.com/envoyproxy/envoy/issues/17771

### CODE REVIEW

File: configs/envoy-demo.yaml [hunk 1]
```diff
@@ -55,3 +55,8 @@ static_resources:
       typed_config:
         "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
         sni: www.envoyproxy.io
+        common_tls_context:
+          validation_context:
+            trusted_ca:
+              filename: certs/upstreamcacert.pem
```
Comment (mattklein123) [inline configs/envoy-demo.yaml:61]: Does this actually work? Doesn't this need to use the system root chain? Same elsewhere.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks for working on this.

> specifically given there's TrustChainVerification ACCEPT_UNTRUSTED I wonder if we should use that for enforcement rather than adding a new boolean. And if we want a boolean if we should name it skip_verification_for_test or if the current name is good or too snarky.

 I would get the thoughts of @envoyproxy/api-shepherds, but I actually prefer an explicit boolean the way you have done it. I actually think this name is good, because there are going to be people that enable this for some reason or another outside of testing whether we like it or not. :)

/wait

**Comment by alyssawilk** [comment]
@lizan or @htuch have alternate preferences?  Otherwise I'll just pick up with Matt's Ok come Monday :-)

File: source/extensions/transport_sockets/tls/context_config_impl.cc [hunk 2]
```diff
@@ -352,6 +370,14 @@ ClientContextConfigImpl::ClientContextConfigImpl(
        config.common_tls_context().tls_certificate_sds_secret_configs().size()) > 1) {
     throw EnvoyException("Multiple TLS certificates are not supported for client contexts");
   }
+  auto optional_validation_context = getValidationContext(config.common_tls_context());
+  if (optional_validation_context.has_value()) {
+    if (!optional_validation_context.value().has_trusted_ca() &&
+        !config.allow_dangerous_tls_without_certificate_validation()) {
+      // FIXME warn for today.
```
Comment (soulxu) [inline source/extensions/transport_sockets/tls/context_config_impl.cc:377]: I guess there should be a runtime flag to control this is a warning message or an exception raising?

I know this is for backward-compatible. But it is strange you have field called 'allow_something', but when it is false, it is just a warning.

**Comment by htuch** [comment]
I think `ACCEPT_UNTRUSTED` is a little different - it says "allow when verification fails if we have a certificate chain"  IIUC. Whereas the new boolean says "if not set, hard fail no matter what if we can't verify for any reason". I think we could change the semantics, but that takes me to my next concern.

Can we safely make this change without this violating xDS backwards compatiblity? If servers expect TLS to work without a CA, are we making an xDS breaking change?

I'm going to argue that "no, it's not sensible to ever expect TLS to be used in this way, it's horribly broken and the intent of the API was not to allow this". However, we have this "Server certificate verification is not enabled by default." comment. It certainly seems to imply a promise here that we support these configurations.

It's then a question of "is this behavior so horribly broken that no sensible server or client implementation would have actually believed this comment?". If so, then I think we can allow the breaking change.

If there are clients or servers that might reasonably depend on this, I don't think we can easily make this change without going through a minor version cycle.

File: configs/envoy-demo.yaml (hunk 1 above)
Comment (lizan) [reply configs/envoy-demo.yaml:61]: also need a `match_subject_alt_names` to be a good demo

**Review by lizan — COMMENTED** [review]
I agree that `ACCEPT_UNTRUSTED` is bit different and that is mostly intended for a server TLS to delegate certificate verification to RBAC/upstream etc, but there's no way to do that in Upstream TLS, thus an additional bool makes sense here.

> It's then a question of "is this behavior so horribly broken that no sensible server or client implementation would have actually believed this comment?". If so, then I think we can allow the breaking change.

I'm not aware of any of this outside testing/experimenting. Could there be some case that only client certificates matter (and ignore whatever server certificate is)?

File: configs/envoy-demo.yaml (hunk 1 above)
Comment (mattklein123) [reply configs/envoy-demo.yaml:61]: Verifying SAN is getting at my point in the linked issue which is that "verifying TLS" is a bit tricky. Obviously verifying the cert against a chain is required, but beyond that there is SAN verification, pinning, etc. and it seems like we might want to enforce different things in different contexts, or for example warn if one of SAN verification, pinning, etc. is not configured.

**Comment by alyssawilk** [comment]
I suspect there are many mesh configurations which will break without setting the new boolean so I would find it plausible that we need to do a bump to make it mandatory.  I think we can change from nothing to logging a warning without going through a version release cycle though.
Even if we're not required to do it now, Do we have an actual process for minor releases yet?

File: configs/envoy-demo.yaml (hunk 1 above)
Comment (alyssawilk) [reply configs/envoy-demo.yaml:61]: I think I'd rather do any changes we're going to do here in one fell swoop, so let's sort out what we think the correct behavior is.
I think making sure we have a valid cert chain is pretty key.  I don't love having to manually  configure match_subject_alt_names.  I would think we'd want to make sure that the host headers sent on a given connection match the SAN for that connection, but I don't see that as a configuration option.

Comment (mattklein123) [reply configs/envoy-demo.yaml:61]: See `auto_san_validation`. The problem here is that this obviously only works for HTTP, and this configuration is generic for all protocols.

This is why I said this is really tricky. :)

I think what you want is something like:
1) Valid cert chain is on by default
2) At least one of a) cert pinning, b) matching a specific SAN, c) auto SAN, etc. is configured.

Comment (mattklein123) [reply configs/envoy-demo.yaml:61]: I did that list of 3 from memory. Please take a look and see if anything else should be in there. cc @ggreenway who knows more about this type of stuff also.

**Comment by htuch** [comment]
Why would mesh configurations break? Typically meshes are strict on mTLS. Agree on the logging front. @adisuissa is driving minor versioning next steps; we're planning on cutting the first minor version early next year.

**Comment by mattklein123** [comment]
> Typically meshes are strict on mTLS.

"meshes" do lots of things. :)

**Comment by htuch** [comment]
Sure, I'm thinking something more like Istio. I don't mind treating this as a runtime migration, rather than API breaking change.

File: configs/envoy-demo.yaml (hunk 1 above)
Comment (alyssawilk) [reply configs/envoy-demo.yaml:61]: coming back to this, I'm a bit concerned that we'll still have a gap here, where if we allow clusters with http configuration to specify auto_san, but then don't have pinned certificates or a default san that we won't actually verify anything for TCP traffic.  Should we require either pinning or a default san just to handle this case?

**Pull request closed without merging**